benchmark/async_hooks/current-resource-vs-destroy.js

```diff
@@ -0,0 +1,165 @@
+'use strict';
+
+const { promisify } = require('util');
+const { readFile } = require('fs');
+const sleep = promisify(setTimeout);
+const read = promisify(readFile);
+const common = require('../common.js');
+const {
+  createHook,
+  currentResource,
+  executionAsyncId
+} = require('async_hooks');
+const { createServer } = require('http');
+
+// configuration for the http server
+// there is no need for parameters in this test
+const connections = 500;
+const path = '/';
+
+const bench = common.createBenchmark(main, {
+  type: ['current-resource', 'destroy'],
+  method: ['callbacks', 'async'],
+  n: [1e6]
+});
+
+function buildCurrentResource(getServe) {
+  const server = createServer(getServe(getCLS, setCLS));
+  const hook = createHook({ init });
+  const cls = Symbol('cls');
+  let closed = false;
+  hook.enable();
+
+  return {
+    server,
+    close
+  };
+
+  function getCLS() {
+    // we need to protect this, as once the hook is
+    // disabled currentResource will return null
+    if (closed) {
+      return;
+    }
+
+    const resource = currentResource();
+    if (!resource[cls]) {
+      return null;
+    }
+    return resource[cls].state;
+  }
+
+  function setCLS(state) {
+    // we need to protect this, as once the hook is
+    // disabled currentResource will return null
+    if (closed) {
+      return;
+    }
+    const resource = currentResource();
+    if (!resource[cls]) {
+      resource[cls] = { state };
+    } else {
+      resource[cls].state = state;
+    }
+  }
+
+  function init(asyncId, type, triggerAsyncId, resource) {
+    if (type === 'TIMERWRAP') return;
+
+    var cr = currentResource();
+    if (cr) {
+      resource[cls] = cr[cls];
```

Comment:
I don't think this necessarily pertinent to the intent of the benchmark here, but I'll call it out nonetheless:  If I'm reading correctly, this results in every resource sharing the same object instance for `cls`, and every `setCLS` call will just overwrite the `state` field.  For correctness here, each resource needs a copy (e.g., `resource[cls] = {...cr[cls]}`), or you need to maintain a "list to the root" (e.g., `resource[cls] = {props:{}, parent: cr[cls]}`.

Reply:
Per conversation, here's an example of where values would be over-written:

```javascript
function f0() {
  setCLS("a");
  setTimeout(function f1() {
     setCLS("b");
     setTimeout(function f2() {
        getCLS(); // expect "b", will get "c"
     }, 100);

  }, 10);

  setTimeout(function f3() => {
      getCLS(); // expect "a", will get "b"
      setCLS("c");
      setTimeout(function f4() {
          getCLS(); // get "c"
      }, 25);
  }, 25);
};
f();
```

Order of execution is `f0`, `f1`, `f3`, `f4`, `f2`. 

But `f0->f1->f2` is one path of the tree, and another indepdent path is `f0->f3->f4`.  Values visible on one path should not be visible on the path.

Reply:
I don't think branch containment really matters for this test. We just need to verify that an object can be passed through the tree via attachment to the current resource in _some_ form.

Reply:
> I don't think branch containment really matters for this test. 

Agreed, but it came up on the diag call & I offered to post example code. :)